Assets/Scripts/Hero/HeroControl.cs

```diff
@@ -1,42 +1,102 @@
 ﻿using UnityEngine;
 using System.Collections;
+using System;
 
 public class HeroControl : MonoBehaviour {
 
+	private static float MOVE_THRESHOLD = 0.19f;
+
 	private Hero hero;
 
-	private bool m_Jump;
+	// Input States
+	private float _moveButtonSpeed = 0.0f;
+	private bool _jumpButtonPressed = false;
+	private bool _crouchButtonPressed = false;
+	private bool _pushButtonPressed = false;
+	private bool _actionButtonPressed = false;
+	private bool _changeHeroButtonPressed = false;
+
 
 	void Awake () {
 		hero = GetComponent<Hero> ();
 	}
 
-	void Update(){
-		if (!m_Jump)
-		{
-			// Read the jump input in Update so button presses aren't missed.
-			m_Jump = Input.GetKeyDown(KeyCode.Space);
-		}
+	private void detectButtonStates(){
+		// Walk
+		_moveButtonSpeed = Input.GetAxis("Horizontal");
+
+		// Change Hero
+		if (Input.GetButtonDown("ChangeHero"))
+			_changeHeroButtonPressed = true;
+		if (Input.GetButtonUp("ChangeHero"))
+			_changeHeroButtonPressed = false;
+
+		// Jump
+		if (Input.GetButtonDown("Jump"))
+			_jumpButtonPressed= true;
+		if (Input.GetButtonUp("Jump"))
+			_jumpButtonPressed = false;
+
+		// Crouch
+		if (Input.GetButtonDown("Crouch"))
+			_crouchButtonPressed= true;
+		if (Input.GetButtonUp("Crouch"))
+			_crouchButtonPressed = false;
 
-        if (Input.GetKeyDown(KeyCode.Tab))
-        {
-            hero.ChangeHero();
-            SoundManager.Instance.SendMessage("PlaySFXSwap");
-        }
+		// Push
+		if (Input.GetButtonDown("Push"))
+			_pushButtonPressed= true;
+		if (Input.GetButtonUp("Push"))
+			_pushButtonPressed = false;
 
-        if (Input.GetButtonDown("Action") && hero.IsActive)
-        {
-            hero.SendMessage("DoAction");
-        }
+		// Action
+		if (Input.GetButtonDown("Action"))
+			_actionButtonPressed= true;
+		if (Input.GetButtonUp("Action"))
+			_actionButtonPressed = false;
 	}
-	
+
+	void Update() {
+		detectButtonStates();
+	}
+
 	void FixedUpdate () {
 
-		if (hero.IsActive)
-			hero.Move (Input.GetAxis ("Horizontal"), Input.GetKey (KeyCode.DownArrow) | Input.GetKey (KeyCode.S), m_Jump);
-		else
-			hero.StopWalk ();
+		if (!hero.IsActive) {
+			hero.StopWalk();
+			return;
+		}
+
+		if (_changeHeroButtonPressed) {
+			hero.ChangeHero();
```

Comment:
Então. Não sei se você testou (eu não ainda), mas acho que to vendo um problema aqui. Enquanto o botão de trocar estiver apertado, o fixed update vai trocar de personagem. Mas o que o cara quer é trocar de hero uma só vez, e aqui esta sendo chamada umas 60 vezes por segundo (que é a taxa de refresh do fixedUpdate).

Você chegou a testar isso no game em si?